src/channel/fuse_pure.rs

```diff
@@ -6,130 +6,87 @@
-        op_size: libc::size_t,
-        userdata: *mut c_void,
-    ) -> *mut c_void;
-    #[cfg(feature = "libfuse3")]
-    pub fn fuse_session_mount(
-        se: *mut c_void, // This argument is really a *fuse_session
-        mountpoint: *const c_char,
-    ) -> c_int;
-    #[cfg(feature = "libfuse3")]
-    // This function's argument is really a *fuse_session
-    pub fn fuse_session_fd(se: *mut c_void) -> c_int;
-    #[cfg(feature = "libfuse3")]
-    // This function's argument is really a *fuse_session
-    pub fn fuse_session_unmount(se: *mut c_void);
-    #[cfg(feature = "libfuse3")]
-    // This function's argument is really a *fuse_session
-    pub fn fuse_session_destroy(se: *mut c_void);
+    }
+}
+
+impl Drop for Mount {
+    fn drop(&mut self) {
+        use std::io::ErrorKind::PermissionDenied;
+        if let Some(sock) = mem::take(&mut self.auto_unmount_socket) {
+            drop(sock);
+            // fusermount in auto-unmount mode, no more work to do.
+            return;
+        }
+        let rc = super::libc_umount(&self.mountpoint);
+        if rc < 0 {
+            let err = io::Error::last_os_error();
+            if err.kind() == PermissionDenied {
+                // Linux always returns EPERM for non-root users.  We have to let the
+                // library go through the setuid-root "fusermount -u" to unmount.
+                fuse_unmount_pure(&self.mountpoint)
+            } else {
+                error!("Unmount failed: {}", err)
+            }
+        }
+    }
 }
 
-#[cfg(not(feature = "libfuse"))]
-pub fn fuse_mount_pure(mountpoint: &OsStr, options: &[MountOption]) -> Result<c_int, io::Error> {
+fn fuse_mount_pure(
+    mountpoint: &OsStr,
+    options: &[MountOption],
+) -> Result<(File, Option<UnixStream>), io::Error> {
     if options.contains(&MountOption::AutoUnmount) {
         // Auto unmount is only supported via fusermount
         return fuse_mount_fusermount(mountpoint, options);
     }
 
     let res = fuse_mount_sys(mountpoint, options)?;
-    if let Some(fd) = res {
-        Ok(fd)
+    if let Some(file) = res {
+        Ok((file, None))
     } else {
         // Retry
         fuse_mount_fusermount(mountpoint, options)
     }
 }
 
-#[cfg(not(feature = "libfuse"))]
-pub fn fuse_unmount_pure(mountpoint: &CStr, fd: c_int) {
-    if fd != -1 {
-        let mut poll_result = libc::pollfd {
-            fd,
-            events: 0,
-            revents: 0,
-        };
-
-        unsafe {
-            let result = libc::poll(&mut poll_result, 1, 0);
-            libc::close(fd);
-            // If the filesystem has already been unmounted, avoid unmounting it again
-            // Unmounting it a second time could cause a race with a newly mounted filesystem
-            // living at the same mountpoint
-            if result > 0 && (poll_result.revents & libc::POLLERR) != 0 {
-                return;
-            }
-        }
-    }
```

Comment:
Why did you remove this? I believe I based this on the behavior of libfuse, which has the same logic

Reply:
This was a mistake, I'd intended to add it back in after the refactoring to do `Arc`. I'll put it back.

Reply:
I've added this back in, but I'm not sure how useful it will be in practice.  I've been unable to get `is_mounted` to return `false` because if it's just detached (lazy unmount) then the `poll` succeeds.